extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,324 @@
+                new ByteBufferAsyncResponseTransformer<>(fragmentSize));
+    }
+
+    private UncheckedDeephavenException handleS3Exception(final Exception e, final String operationDescription) {
+        if (e instanceof InterruptedException) {
+            Thread.currentThread().interrupt();
+            return new UncheckedDeephavenException("Thread interrupted while " + operationDescription, e);
+        } else if (e instanceof ExecutionException) {
+            return new UncheckedDeephavenException("Execution exception occurred while " + operationDescription, e);
+        } else if (e instanceof TimeoutException) {
+            return new UncheckedDeephavenException("Operation timeout while " + operationDescription + " after waiting "
+                    + "for duration " + readTimeout, e);
+        } else if (e instanceof CancellationException) {
+            return new UncheckedDeephavenException("Cancelled an operation while " + operationDescription, e);
+        }
+        throw new UncheckedDeephavenException("Exception caught while " + operationDescription, e);
+    }
+
+    @Override
+    public int write(final ByteBuffer src) throws IOException {
+        throw new NonWritableChannelException();
+    }
+
+    @Override
+    public long position() throws ClosedChannelException {
+        final long localPosition = position;
+        checkClosed(localPosition);
+        return localPosition;
+    }
+
+    @Override
+    public SeekableByteChannel position(final long newPosition) throws ClosedChannelException {
+        if (newPosition < 0) {
+            throw new IllegalArgumentException("newPosition cannot be < 0, provided newPosition=" + newPosition);
+        }
+        checkClosed(position);
+        position = newPosition;
+        return this;
+    }
+
+    @Override
+    public long size() throws ClosedChannelException {
+        checkClosed(position);
+        if (size < 0) {
+            populateSize();
+        }
+        return size;
+    }
+
+    private void populateSize() {
+        if (context.getSize() < 0) {
+            // Fetch the size of the file on the first read using a blocking HEAD request
+            final HeadObjectResponse headObjectResponse;
+            try {
+                headObjectResponse = s3AsyncClient.headObject(builder -> builder.bucket(bucket).key(key))
+                        .get(readTimeout.toNanos(), TimeUnit.NANOSECONDS);
+            } catch (final InterruptedException | ExecutionException | TimeoutException | CancellationException e) {
+                final String operation = "fetching HEAD for file " + key + " in S3 bucket " + bucket;
+                throw handleS3Exception(e, operation);
+            }
+            context.setSize(headObjectResponse.contentLength());
+        }
+        this.size = context.getSize();
+        this.numFragmentsInObject = (int) ((size + fragmentSize - 1) / fragmentSize); // = ceil(size / fragmentSize)
```

Comment:
We only need to set these once, right? Structurally, I would just return from this method if `this.size != -1`. (Technically, an s3 file could be empty.)